firewood/src/manager.rs

```diff
@@ -59,129 +73,117 @@ pub(crate) enum RevisionManagerError {
+    /// 7. Proposal Cleanup.
+    ///    Any other proposals that have this proposal as a parent should be reparented to the committed version.
+    /// 8. Revision reaping.
+    ///    If more than the configurable number of revisions is available, the oldest revision can be forgotten.
+    pub fn commit(
+        &mut self,
+        proposal: Arc<NodeStore<ImmutableProposal, FileBacked>>,
+    ) -> Result<(), RevisionManagerError> {
+        // 1. Commit check
+        let current_revision = self.current_revision();
+        if !proposal
+            .kind
+            .parent_is(&current_revision.kind.as_nodestore_parent())
+        {
+            return Err(RevisionManagerError::NotLatest);
+        }
+        // 2. Persist delete list
+        // TODO
+
+        // 3. Set last committed revision
+        let committed: Arc<NodeStore<Committed, FileBacked>> = proposal.as_committed().into();
+        self.historical.push_back(committed.clone());
+        if let Some(hash) = committed.kind.root_hash() {
+            self.by_hash.insert(hash, committed.clone());
+        }
+        // TODO: We could allow other commits to start here using the pending list
+
+        // 4. Free list flush
+        proposal.flush_freelist()?;
+
+        // 5. Node flush
+        proposal.flush_nodes()?;
+
+        // 6. Root move
+        proposal.flush_header()?;
+
+        // 7. Proposal Cleanup
+        self.proposals.retain(|p| {
+            // TODO: reparent proposals; this needs a lock on the parent element of immutable proposals
+            // if p
+            //     .kind
+            //     .parent_is(&proposal.kind.as_nodestore_parent())
+            // {
+            //     p.kind.reparent_to(&committed.kind.as_nodestore_parent());
+            // }
+            !Arc::ptr_eq(&proposal, p)
+        });
+
+        Ok(())
     }
 }
 
 pub type NewProposalError = (); // TODO implement
 
 impl RevisionManager {
-    // TODO fix this or remove it. It should take in a proposal.
-    pub fn add_proposal(&mut self, _proposal: ()) {
-        todo!()
-        // self.proposals.push(proposal);
+    pub fn add_proposal(&mut self, proposal: Arc<NodeStore<ImmutableProposal, FileBacked>>) {
+        self.proposals.push(proposal);
+    }
+
+    pub fn revision(
+        &self,
+        root_hash: HashKey,
+    ) -> Result<Arc<NodeStore<Committed, FileBacked>>, RevisionManagerError> {
+        self.by_hash
+            .get(&root_hash)
+            .cloned()
+            .ok_or(RevisionManagerError::IO(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                "Revision not found",
+            )))
     }
 
-    pub fn revision(&self, _root_hash: HashKey) -> Result<(), RevisionManagerError> {
-        todo!()
+    pub fn root_hash(&self) -> Result<Option<HashKey>, RevisionManagerError> {
+        self.current_revision()
+            .kind
+            .root_hash()
+            .map(Option::Some)
```

Comment:
IIRC this can only happens during some tests, otherwise you're right, we always have a valid latest revision. I wish there was a type that didn't allow for an empty list.

I will fix this in a followup.